Sources/Structs/Weak.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  Created by Murat Yilmaz on 19.05.18.
+//  Copyright © 2018 Flinesoft. All rights reserved.
+//
+
+import Foundation
+
+/// A wrapper for storing weak references to a `Wrapped` instance.
+public struct Weak<Wrapped>: ExpressibleByNilLiteral where Wrapped: AnyObject {
+    /// The value of `Wrapped` stored as weak reference
+    public weak var value: Wrapped?
+
+    /// Creates an instance that stores the given value.
+    public init(_ value: Wrapped) {
+        self.value = value
+    }
+
+    /// Evaluates the given closure when this `Weak` instance is not `nil`,
+    /// passing the value as a parameter.
+    ///
+    /// - Parameter transform: A closure that takes the unwrapped value
+    ///   of the instance.
+    /// - Returns: The result of the given closure. If this instance is `nil`,
+    ///   returns `nil`.
+    public func map<U>(_ transform: (Wrapped) throws -> U) rethrows -> U? {
```

Comment:
I'm not sure about the `map` and `flatMap` here. Shouldn't we remove those and only include `compactMap` which is more expressive about unwrapping the value? I mean, even `flatMap` was deprecated in favor of `compactMap` in Swift 4.1, so we should resemble that thinking here, too, no?

Reply:
By the way, I'm assuming that this works nicely with arrays of `Weak`.

Reply:
Hmm ... I will need to test this out on the usage side. We should get the API naming right here.